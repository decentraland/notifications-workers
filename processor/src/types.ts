import {
  IBaseComponent,
  IConfigComponent,
  IFetchComponent,
  IHttpServerComponent,
  ILoggerComponent,
  IMetricsComponent
} from '@well-known-components/interfaces'
import { metricDeclarations } from '@well-known-components/logger'
import { IPgComponent } from '@well-known-components/pg-component'
import { DecentralandSignatureContext } from '@dcl/platform-crypto-middleware'
import { ISubgraphComponent } from '@well-known-components/thegraph-component'
import { DbComponent } from './adapters/db'

export type AppComponents = {
  config: IConfigComponent
  logs: ILoggerComponent
  server: IHttpServerComponent<GlobalContext>
  metrics: IMetricsComponent<keyof typeof metricDeclarations>
  pg: IPgComponent
  db: DbComponent
  statusChecks: IBaseComponent
  producerRegistry: IProducerRegistry
  l2CollectionsSubGraph: ISubgraphComponent
  marketplaceSubGraph: ISubgraphComponent
  rentalsSubGraph: ISubgraphComponent
  landManagerSubGraph: ISubgraphComponent
}

// this type simplifies the typings of http handlers
export type HandlerContextWithPath<
  ComponentNames extends keyof AppComponents,
  Path extends string = any
> = IHttpServerComponent.PathAwareContext<
  IHttpServerComponent.DefaultContext<{
    components: Pick<AppComponents, ComponentNames>
  }>,
  Path
> &
  DecentralandSignatureContext<any>

export type GlobalContext = {
  components: AppComponents
}

// components used in tests
export type TestComponents = AppComponents & {
  // A fetch component that only hits the test server
  localFetch: IFetchComponent
}

export type INotificationProducer = {
  start: () => Promise<void>
  notificationType: () => string
  runProducerSinceDate(date: number): Promise<void>
}

export type INotificationGenerator = {
  run(since: number): Promise<INotificationProducerResult>
  notificationType: string
}

export type IProducerRegistry = IBaseComponent & {
  addProducer: (producer: INotificationProducer) => void
  getProducer: (notificationType: string) => INotificationProducer
}

export type NotificationRecord = {
  eventKey: string
  type: string
  address: string
  metadata: any
  timestamp: number
}

export type INotificationProducerResult = {
  notificationType: string
  records: NotificationRecord[]
  lastRun: number
}
