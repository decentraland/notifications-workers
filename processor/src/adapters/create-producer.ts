import { AppComponents, INotificationGenerator, INotificationProducer } from '../types'
import { CronJob } from 'cron'

export async function createProducer(
  components: Pick<AppComponents, 'logs' | 'db'>,
  producer: INotificationGenerator
): Promise<INotificationProducer> {
  const { logs, db } = components
  const logger = logs.getLogger(`producer-${producer.notificationType}`)

  let lastSuccessfulRun: Date | undefined

  async function runProducer() {
    if (!lastSuccessfulRun) {
      lastSuccessfulRun = new Date(await db.fetchLastUpdateForNotificationType(producer.notificationType))
    }

    logger.info('Checking for updates since ' + lastSuccessfulRun.toISOString())

    const produced = await producer.run(lastSuccessfulRun)
    await db.insertNotifications(produced.records)
    await db.updateLastUpdateForNotificationType(produced.notificationType, produced.lastRun)
    lastSuccessfulRun = produced.lastRun
    logger.info(`Created ${produced.records.length} new notifications`)
  }

  async function start(): Promise<void> {
    logger.info(`Scheduling producer for ${producer.notificationType}.`)

    const job = new CronJob(
      '0 * * * * *',
      async function () {
        try {
          await runProducer()
        } catch (e: any) {
          logger.error(`Couldn't run producer: ${e.message}`)
        }
      },
      null,
      false,
      'UCT'
    )
    job.start()
  }

  return {
    start,
    notificationType: () => producer.notificationType,
    setLastSuccessfulRun: (date: Date) => {
      lastSuccessfulRun = date
    }
  }
}
