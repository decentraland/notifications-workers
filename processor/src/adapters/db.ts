import SQL from 'sql-template-strings'
import { AppComponents, NotificationRecord } from '../types'

export type DbComponent = {
  fetchLastUpdateForNotificationType(notificationType: string): Promise<number>
  updateLastUpdateForNotificationType(notificationType: string, timestamp: number): Promise<void>
  insertNotifications(notificationRecord: NotificationRecord[]): Promise<void>
}

export function createDbComponent({ pg }: Pick<AppComponents, 'pg' | 'logs'>): DbComponent {
  async function fetchLastUpdateForNotificationType(notificationType: string): Promise<number> {
    const result = await pg.query<{ last_successful_run_at: number }>(SQL`
        SELECT *
        FROM cursors
        WHERE id = ${notificationType};
    `)
    if (result.rowCount === 0) {
      return Date.now()
    }

    return result.rows[0].last_successful_run_at
  }

  async function updateLastUpdateForNotificationType(notificationType: string, timestamp: number): Promise<void> {
    const query = SQL`
        INSERT INTO cursors (id, last_successful_run_at, created_at, updated_at)
        VALUES (${notificationType}, ${timestamp}, ${Date.now()}, ${Date.now()})
        ON CONFLICT (id) DO UPDATE
        SET last_successful_run_at = ${timestamp},
            updated_at             = ${Date.now()};
    `

    await pg.query<any>(query)
  }

  async function insertNotifications(notificationRecords: NotificationRecord[]) {
    if (notificationRecords.length === 0) {
      return
    }

    for (const notificationRecord of notificationRecords) {
      const buildQuery = SQL`
          INSERT INTO notifications (event_key, type, address, metadata, timestamp, read_at, created_at, updated_at)
          VALUES (${notificationRecord.eventKey},
                  ${notificationRecord.type},
                  ${notificationRecord.address?.toLowerCase() || null},
                  ${notificationRecord.metadata}::jsonb,
                  ${notificationRecord.timestamp},
                  NULL,
                  ${Date.now()},
                  ${Date.now()})
          ON CONFLICT (event_key, type, address) DO UPDATE
              SET metadata   = ${notificationRecord.metadata}::jsonb,
                  timestamp  = ${notificationRecord.timestamp},
                  updated_at = ${Date.now()};
      `
      await pg.query(buildQuery)
    }
  }

  return {
    fetchLastUpdateForNotificationType,
    updateLastUpdateForNotificationType,
    insertNotifications
  }
}
