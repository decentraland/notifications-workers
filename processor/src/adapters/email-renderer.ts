import * as fs from 'node:fs'
import * as path from 'path'
import handlebars from 'handlebars'
import { NotificationType } from '@dcl/schemas'
import { Email } from '@notifications/common'
import { AppComponents, IEmailRenderer, NotificationRecord } from '../types'
import { formatMana } from '../logic/utils'

const nonTransformer = (notification: NotificationRecord): NotificationRecord => notification

const formatPriceAsMana = (notification: NotificationRecord) => ({
  ...notification,
  metadata: {
    ...notification.metadata,
    price: formatMana(notification.metadata.price)
  }
})

const formatRoyaltiesCutAsMana = (notification: NotificationRecord) => ({
  ...notification,
  metadata: {
    ...notification.metadata,
    royaltiesCut: formatMana(notification.metadata.royaltiesCut)
  }
})

const transformers: Partial<Record<NotificationType, (notification: NotificationRecord) => NotificationRecord>> = {
  [NotificationType.BID_ACCEPTED]: formatPriceAsMana,
  [NotificationType.BID_RECEIVED]: formatPriceAsMana,
  [NotificationType.ROYALTIES_EARNED]: formatRoyaltiesCutAsMana
}

function loadTemplates(part: string) {
  return Object.values(NotificationType)
    .map((type) => fs.readFileSync(path.join(__dirname, `email-templates/${type}.${part}.handlebars`), 'utf8'))
    .reduce(
      (acc, template, index) => {
        acc[Object.values(NotificationType)[index]] = handlebars.compile(template, {})
        return acc
      },
      {} as Record<NotificationType, HandlebarsTemplateDelegate>
    )
}

export async function createRenderer(components: Pick<AppComponents, 'subscriptionService'>): Promise<IEmailRenderer> {
  const { subscriptionService } = components

  const contentTemplates = loadTemplates('content')
  const subjectTemplates = loadTemplates('subject')

  async function renderEmail(notification: NotificationRecord): Promise<Email> {
    const subscription = await subscriptionService.findSubscriptionForAddress(notification.address)
    const transformed = (transformers[notification.type] || nonTransformer)(notification)
    return {
      to: subscription?.email || '',
      subject: subjectTemplates[notification.type](transformed),
      content: contentTemplates[notification.type](transformed)
    }
  }

  return {
    renderEmail
  }
}
