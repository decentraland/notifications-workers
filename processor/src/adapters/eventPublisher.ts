import { PublishCommand, SNSClient } from '@aws-sdk/client-sns'
import { EventNotification } from '../event.types'
import { AppComponents, IEventPublisher } from '../types'

export async function createEventPublisher({ config }: Pick<AppComponents, 'config'>): Promise<IEventPublisher> {
  const snsArn = await config.requireString('AWS_SNS_ARN')
  const optionalEndpoint = await config.getString('AWS_SNS_ENDPOINT')

  const client = new SNSClient({
    endpoint: optionalEndpoint
  })

  async function publishMessage(event: EventNotification): Promise<string | undefined> {
    const { MessageId } = await client.send(
      new PublishCommand({
        TopicArn: snsArn,
        Message: JSON.stringify(event)
      })
    )

    return MessageId
  }

  return { publishMessage }
}
