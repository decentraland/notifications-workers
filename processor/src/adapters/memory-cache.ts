import { IMemoryCache } from '../types'

export function createInMemoryCacheComponent(): IMemoryCache {
  const cache = new Map<string, any>()

  function get(key: string): any {
    return cache.get(key)
  }

  function set(key: string, value: any): void {
    cache.set(key, value)
  }

  return {
    get,
    set
  }
}
