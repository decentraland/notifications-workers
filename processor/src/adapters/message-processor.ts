import { AppComponents, IMessageProcessor } from '../types'
import { sleep } from '../logic/utils'
import { NotificationRecord } from '@notifications/common'

export function createMessageProcessor({
  logs,
  queueConsumer,
  notificationsService,
  eventParser
}: Pick<AppComponents, 'logs' | 'queueConsumer' | 'notificationsService' | 'eventParser'>): IMessageProcessor {
  const logger = logs.getLogger('messages-consumer')
  let isRunning = false

  function parseMessageToNotification(message: string): NotificationRecord | undefined {
    try {
      logger.info('Pulled message from queue', { message })
      const parsedMessage = JSON.parse(message)
      return eventParser.parseToNotification(parsedMessage)
    } catch (error: any) {
      logger.error(`Failed while parsing message from queue: ${error?.message || 'Unexpected failure'}`)
      return undefined
    }
  }

  async function start() {
    logger.info('Starting to listen messages from queue')
    isRunning = true
    while (isRunning) {
      const messages = await queueConsumer.receiveMessages()

      if (messages.length === 0) {
        logger.info('No messages found in queue, waiting 10 seconds to check again')
        await sleep(10 * 1000)
        continue
      }

      for (const message of messages) {
        const { Body, ReceiptHandle } = message
        logger.info('Pulled message from queue', { message: Body! })
        const notification: NotificationRecord | undefined = parseMessageToNotification(Body!)

        if (!notification) {
          await queueConsumer.deleteMessage(ReceiptHandle!)
          continue
        }

        try {
          await notificationsService.saveNotifications([notification])
          logger.info(`Created a new ${notification.type} notification.`)
        } catch (error: any) {
          // TODO: handle retries and DLQ
          logger.error(`Failed while processing event notification: ${error?.message || 'Unexpected failure'}`)
        } finally {
          await queueConsumer.deleteMessage(ReceiptHandle!)
        }
      }
    }
  }

  async function stop() {
    isRunning = false
  }

  return { start, stop }
}
