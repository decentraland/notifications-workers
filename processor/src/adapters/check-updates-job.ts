import { AppComponents, INotificationProducer } from '../types'
import { itemSoldProducer } from '../logic/producers/item-sold'
import { IBaseComponent } from '@well-known-components/interfaces'
import { createProducer } from '../logic/producers/create-producer'

export async function createCheckUpdatesJob(
  components: Pick<AppComponents, 'config' | 'db' | 'logs' | 'l2CollectionsSubGraph'>
): Promise<IBaseComponent> {
  const producers: INotificationProducer[] = []

  async function start(): Promise<void> {
    // Add more producers here
    producers.push(await itemSoldProducer(components))

    for (const producer of producers) {
      const p = await createProducer(components, producer)
      await p.run()
    }
  }

  return {
    start
  }
}
