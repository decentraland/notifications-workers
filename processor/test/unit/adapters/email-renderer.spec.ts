import { createRenderer } from '../../../src/adapters/email-renderer'
import { IEmailRenderer, NotificationRecord } from '../../../src/types'
import { NotificationType } from '@dcl/schemas'

describe('email rendering tests', () => {
  let renderer: IEmailRenderer

  const notifications: Record<NotificationType, NotificationRecord> = {
    [NotificationType.BID_ACCEPTED]: {
      type: NotificationType.BID_ACCEPTED,
      address: '0x1234567890ABCDEF1234567890ABCDEF12345678',
      metadata: {
        link: 'https://decentraland.org/marketplace/contracts/0x557539e7792dc12a0555f5ff02d6ec0f0bc88e09/tokens/21',
        image:
          'https://peer.decentraland.org/lambdas/collections/contents/urn:decentraland:matic:collections-v2:0x557539e7792dc12a0555f5ff02d6ec0f0bc88e09:0/thumbnail',
        price: '1500000000000000000',
        title: 'Bid Accepted',
        rarity: 'legendary',
        seller: '0x072f890d413c5b37b2ee7661929dfb681abc6673',
        network: 'polygon',
        nftName: 'Netgate male',
        category: 'wearable',
        description: 'Your bid for 1.50 MANA for this Netgate male was accepted.'
      },
      timestamp: Date.now(),
      eventKey: '123'
    },
    [NotificationType.BID_RECEIVED]: {
      type: NotificationType.BID_RECEIVED,
      address: '0x1234567890ABCDEF1234567890ABCDEF12345678',
      metadata: {
        link: 'https://decentraland.org/marketplace/account?assetType=nft&section=bids',
        image:
          'https://peer.decentraland.org/lambdas/collections/contents/urn:decentraland:matic:collections-v2:0xf6ed0f2eea302dbd77482bd4771645ee142ecf7c:1/thumbnail',
        price: '1500000000000000000',
        title: 'Bid Received',
        rarity: 'legendary',
        seller: '0x460b8af3f189aa7f396b4ef696f1218f0fcf7f28',
        network: 'polygon',
        nftName: 'Space Jams',
        category: 'wearable',
        description: 'You received a bid of 1.50 MANA for this Space Jams.'
      },
      timestamp: Date.now(),
      eventKey: '123'
    },
    [NotificationType.EVENTS_STARTS_SOON]: {
      type: NotificationType.EVENTS_STARTS_SOON,
      address: '0x1234567890ABCDEF1234567890ABCDEF12345678',
      metadata: {
        link: 'https://play.decentraland.org/?position=-99%2C35&realm=main',
        name: 'UNMONDAY YOURSELF',
        image: 'https://events-assets-099ac00.decentraland.org/poster/215052d46626a2a8.png',
        title: 'Event starts in an hour',
        endsAt: '2024-05-20T21:00:00.000Z',
        startsAt: '2024-03-18T19:00:00.000Z',
        description: 'The event UNMONDAY  YOURSELF starts in an hour.'
      },
      timestamp: Date.now(),
      eventKey: '123'
    },
    [NotificationType.EVENTS_STARTED]: {
      type: NotificationType.EVENTS_STARTED,
      address: '0x1234567890ABCDEF1234567890ABCDEF12345678',
      metadata: {
        link: 'https://play.decentraland.org/?position=14%2C100&realm=main',
        name: 'Live DJ Music and Dancing in the Unity Café',
        image: 'https://events-assets-099ac00.decentraland.org/poster/3b8e6fc465ee4488.png',
        title: 'Event started',
        description: 'The event Live DJ Music and Dancing in the Unity Café has begun!'
      },
      timestamp: Date.now(),
      eventKey: '123'
    },
    [NotificationType.GOVERNANCE_ANNOUNCEMENT]: {
      type: NotificationType.GOVERNANCE_ANNOUNCEMENT,
      address: null,
      metadata: {
        title: 'Some title',
        description: 'Some content',
        link: 'https://governance.decentraland.org/some-link'
      },
      timestamp: Date.now(),
      eventKey: '123'
    },
    [NotificationType.GOVERNANCE_AUTHORED_PROPOSAL_FINISHED]: {
      type: NotificationType.GOVERNANCE_AUTHORED_PROPOSAL_FINISHED,
      address: '0x1234567890ABCDEF1234567890ABCDEF12345678',
      metadata: {
        link: 'https://governance.decentraland.org/proposal/?id=c9564934-1f5e-44a3-836e-c0e94c1f593f',
        title: 'Voting ended on your proposal Limit Grants Re-Submissions',
        proposalId: 'c9564934-1f5e-44a3-836e-c0e94c1f593f',
        description: 'The votes are in! Find out the outcome of the voting on your proposal now',
        proposalTitle: 'Limit Grants Re-Submissions'
      },
      timestamp: Date.now(),
      eventKey: '123'
    },
    [NotificationType.GOVERNANCE_COAUTHOR_REQUESTED]: undefined,
    [NotificationType.GOVERNANCE_NEW_COMMENT_ON_PROJECT_UPDATE]: undefined,
    [NotificationType.GOVERNANCE_NEW_COMMENT_ON_PROPOSAL]: undefined,
    [NotificationType.GOVERNANCE_PITCH_PASSED]: undefined,
    [NotificationType.GOVERNANCE_PROPOSAL_ENACTED]: undefined,
    [NotificationType.GOVERNANCE_TENDER_PASSED]: undefined,
    [NotificationType.GOVERNANCE_VOTING_ENDED_VOTER]: undefined,
    [NotificationType.ITEM_SOLD]: undefined,
    [NotificationType.LAND_RENTAL_ENDED]: undefined,
    [NotificationType.LAND_RENTED]: undefined,
    [NotificationType.REWARD_ASSIGNED]: undefined,
    [NotificationType.ROYALTIES_EARNED]: undefined,
    [NotificationType.WORLDS_ACCESS_RESTORED]: undefined,
    [NotificationType.WORLDS_ACCESS_RESTRICTED]: undefined,
    [NotificationType.WORLDS_MISSING_RESOURCES]: undefined,
    [NotificationType.WORLDS_PERMISSION_GRANTED]: undefined,
    [NotificationType.WORLDS_PERMISSION_REVOKED]: undefined
  }

  beforeAll(() => {
    renderer = createRenderer()
  })

  const cases = Object.keys(notifications).map((type) => [type, notifications[type]])

  test('Examples match the type of their keys', () => {
    Object.keys(notifications).forEach((type) => {
      expect(notifications[type].type).toBe(type)
    })
  })

  test.each(cases)(`rendering %s`, async (type: NotificationType, notification: NotificationRecord) => {
    expect(renderer.renderEmail(notification)).toMatchSnapshot()
  })
})
