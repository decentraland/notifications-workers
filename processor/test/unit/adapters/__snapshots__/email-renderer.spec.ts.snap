// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`email rendering tests rendering bid_accepted 1`] = `
{
  "actionButtonText": "Check it Out",
  "actionButtonUrl": "https://decentraland.org/marketplace/contracts/0x557539e7792dc12a0555f5ff02d6ec0f0bc88e09/tokens/21",
  "content": "<p>Your offer of 1.50 MANA was accepted for Netgate male.</p>
",
  "subject": "Your offer was accepted",
  "to": "email@example.com",
}
`;

exports[`email rendering tests rendering bid_received 1`] = `
{
  "actionButtonText": "Manage Your Bids",
  "actionButtonUrl": "https://decentraland.org/marketplace/account?assetType=nft&section=bids",
  "content": "<p>You've received an offer of 1.50 MANA for Space Jams.</p>
",
  "subject": "You've Received a Bid",
  "to": "email@example.com",
}
`;

exports[`email rendering tests rendering events_started 1`] = `
{
  "actionButtonText": "Jump In",
  "actionButtonUrl": "https://play.decentraland.org/?position=14%2C100&realm=main",
  "content": "<p>The event <a href="https://play.decentraland.org/?position=14%2C100&realm=main">Live DJ Music and Dancing in the Unity Café</a> has begun!</p>
",
  "subject": "Jump In - Your Event Has Started!",
  "to": "email@example.com",
}
`;

exports[`email rendering tests rendering events_starts_soon 1`] = `
{
  "actionButtonText": "Go to Event Page",
  "actionButtonUrl": "https://play.decentraland.org/?position=-99%2C35&realm=main",
  "content": "<p>The event <a href="https://play.decentraland.org/?position=-99%2C35&realm=main">UNMONDAY YOURSELF</a> starts in an hour.</p>
",
  "subject": "An Event You've RSVP'd to Starts Soon",
  "to": "email@example.com",
}
`;

exports[`email rendering tests rendering governance_announcement 1`] = `
{
  "content": "<p style="font-weight: 600; font-size: 14px; margin-bottom: 0;">Some title</p>
<p style="display: inline; color: #a09ba8; margin: 4px 0; gap: 5px; flex-wrap: wrap; align-items: center;">
    <a href="https://governance.decentraland.org/some-link">
        Some content
    </a>
</p>
",
  "subject": "A DAO Announcement Has Been Released",
  "to": "email@example.com",
}
`;

exports[`email rendering tests rendering governance_authored_proposal_finished 1`] = `
{
  "actionButtonText": "See the Results",
  "actionButtonUrl": "https://governance.decentraland.org/proposal/?id=c9564934-1f5e-44a3-836e-c0e94c1f593f",
  "content": "<p>The votes are in! Find out the outcome of the voting on your proposal Limit Grants Re-Submissions.</p>
",
  "subject": "Voting Has Ended on Your Proposal",
  "to": "email@example.com",
}
`;

exports[`email rendering tests rendering governance_coauthor_requested 1`] = `
{
  "actionButtonText": "Respond to the Request",
  "actionButtonUrl": "https://governance.decentraland.org/proposal/?id=cacb1ce3-b5e9-4133-901a-cb787f43871f",
  "content": "<p>You've been invited to collaborate as a co-author on the proposal DCL Community Video Library.</p>
",
  "subject": "You've Received a Request to Co-Author a Proposal",
  "to": "email@example.com",
}
`;

exports[`email rendering tests rendering governance_new_comment_on_project_update 1`] = `
{
  "actionButtonText": "Read the Comment",
  "actionButtonUrl": "https://governance.decentraland.org/update/?id=a38ec7d7-7eb2-4f9a-b9c8-363785fdcbd6&proposalId=1b8eef75-add3-49d6-9181-3061daaf22ec",
  "content": "<p>Someone's commented on your project update for Duel Arena P2E Continuum, Refinement &amp; Game Expansion [ Resubmission ].</p>",
  "subject": "There's a New Comment on Your Project Update",
  "to": "email@example.com",
}
`;

exports[`email rendering tests rendering governance_new_comment_on_proposal 1`] = `
{
  "actionButtonText": "Read the Comment",
  "actionButtonUrl": "https://governance.decentraland.org/proposal/?id=578d3d3e-9fd2-48cb-bea3-ddc6fb5ab1dd",
  "content": "<p>Someone's commented on your proposal Decentraland No-Code UI Design Studio.</p>
",
  "subject": "There's a New Comment on Your Proposal",
  "to": "email@example.com",
}
`;

exports[`email rendering tests rendering governance_pitch_passed 1`] = `
{
  "actionButtonText": "See the Pitch",
  "actionButtonUrl": "https://governance.decentraland.org/proposal/?id=c9564934-1f5e-44a3-836e-c0e94c1f593f",
  "content": "<p>The pitch Limit Grants Re-Submissions has passed and can now receive tenders. Help advance this idea by proposing potential solutions.</p>",
  "subject": "A Pitch You Voted on Has Passed",
  "to": "email@example.com",
}
`;

exports[`email rendering tests rendering governance_proposal_enacted 1`] = `
{
  "actionButtonText": "See Proposal",
  "actionButtonUrl": "https://governance.decentraland.org/proposal/?id=578d3d3e-9fd2-48cb-bea3-ddc6fb5ab1dd",
  "content": "<p>Congratulations! Your Proposal Decentraland No-Code UI Design Studio has been enacted!</p>",
  "subject": "Your Proposal's Been Enacted!",
  "to": "email@example.com",
}
`;

exports[`email rendering tests rendering governance_tender_passed 1`] = `
{
  "actionButtonText": "See the Tender",
  "actionButtonUrl": "https://governance.decentraland.org/proposal/?id=c9564934-1f5e-44a3-836e-c0e94c1f593f",
  "content": "The tender you voted on Limit Grants Re-Submissions has passed and can now receive bids. If you think you can tackle this solution, make a bid and propose a project!",
  "subject": "A Tender You Voted on Has Passed",
  "to": "email@example.com",
}
`;

exports[`email rendering tests rendering governance_voting_ended_voter 1`] = `
{
  "actionButtonText": "See the Results",
  "actionButtonUrl": "https://governance.decentraland.org/proposal/?id=c9564934-1f5e-44a3-836e-c0e94c1f593f",
  "content": "<p>Voting has ended on the proposal Limit Grants Re-Submissions, see the outcome of your vote!</p>
",
  "subject": "Voting Has Ended on a Proposal You Voted On",
  "to": "email@example.com",
}
`;

exports[`email rendering tests rendering item_sold 1`] = `
{
  "actionButtonText": "Go to Wallet",
  "actionButtonUrl": "https://decentraland.org/marketplace/contracts/0xca520eea5aadff51b48d9e9b3038001a751139ca/tokens/180",
  "content": "<p>Someone just bought your item HeartPrint Shoes. Congratulations!</p>
",
  "subject": "You’ve Sold an Item!",
  "to": "email@example.com",
}
`;

exports[`email rendering tests rendering rental_ended 1`] = `
{
  "content": "The rental of your <a style="color: #ff2d55; text-decoration: underline;" href="https://decentraland.org/marketplace/contracts/0x959e104e1a4db6317fa58f8295f586e1a978c297/tokens/1249/manage">ESTATE</a> has ended.
",
  "subject": "",
  "to": "email@example.com",
}
`;

exports[`email rendering tests rendering rental_started 1`] = `
{
  "content": "Your <a style="color: #ff2d55; text-decoration: underline;" href="https://decentraland.org/marketplace/contracts/0x959e104e1a4db6317fa58f8295f586e1a978c297/tokens/5051/manage">ESTATE</a> was rented by 0x951a6698f644a9b1c634abc416b9d9902aaf3db9.
",
  "subject": "",
  "to": "email@example.com",
}
`;

exports[`email rendering tests rendering reward_assignment 1`] = `
{
  "content": "This MESHfair 2024 Halo is already in your backpack.
",
  "subject": "",
  "to": "email@example.com",
}
`;

exports[`email rendering tests rendering royalties_earned 1`] = `
{
  "content": "You've earned 0.90 MANA in royalties for <a style="color: #ff2d55; text-decoration: underline;" href="https://decentraland.org/marketplace/contracts/0x08cbc78c1b2e2eea7627c39c8adf660d52e3d82c/tokens/945">Blue santa hat</a>.
",
  "subject": "",
  "to": "email@example.com",
}
`;

exports[`email rendering tests rendering worlds_access_restored 1`] = `
{
  "content": "Access to your <a style="color: #ff2d55; text-decoration: underline;" href="https://decentraland.org/builder/worlds?tab=dcl">Worlds</a> has been restored.
",
  "subject": "",
  "to": "email@example.com",
}
`;

exports[`email rendering tests rendering worlds_access_restricted 1`] = `
{
  "content": "Access to your Worlds has been restricted due to <a style="color: #ff2d55; text-decoration: underline;" href="https://decentraland.org/builder/worlds?tab=dcl">insufficient resources</a>.
",
  "subject": "",
  "to": "email@example.com",
}
`;

exports[`email rendering tests rendering worlds_missing_resources 1`] = `
{
  "content": "Access to your Worlds at risk in 48h. <a style="color: #ff2d55; text-decoration: underline;" href="https://decentraland.org/builder/worlds?tab=dcl">Rectify now</a> to prevent disruption.
)
",
  "subject": "",
  "to": "email@example.com",
}
`;

exports[`email rendering tests rendering worlds_permission_granted 1`] = `
{
  "content": "You have been granted access permission for world el.dcl.eth.
",
  "subject": "",
  "to": "email@example.com",
}
`;

exports[`email rendering tests rendering worlds_permission_revoked 1`] = `
{
  "content": "Your streaming permission for world el.dcl.eth has been revoked.
",
  "subject": "",
  "to": "email@example.com",
}
`;
